### PR TITLE
Fixed to use converter

### PR DESCRIPTION
**Pull request by kan-bayashi**
unify the data converter with TTS one #334 

- cleaned up `nets/e2e_asr_th.py`
    - improved variable name consistency
    - removed `linear_tensor` (no need to use) 
    - removed unnecessary lines
    - changed input format
    - moved some functions in functions
- cleaned up `nets/e2e_asr.py`
    - removed unnecessary lines
    - changed input format
    - moved some functions in functions
- cleaned up `asr/asr_pytorch.py`
    - renamed `Updater` and `Evaluator`
    - use new converter
    - replaced `.cuda()` with `.to(device)` (pytorch official recommendation)
- cleand up `asr/asr_chainer.py`
    - renamed `Updater` and `Evaluator`
    - use new converter
- updated test codes
    - make it compatible with new input format of e2e model
- updated `PlotAttentionReport`
    - made the use more clearer

Now testing
- [x] chainer single / wsj 
- [x] chainer multi / wsj
- [x] pytorch single / wsj
- [x] pytorch multi / wsj 
- [x] pytorch single / ljspeech
- [x] pytorch multi /ljspeech

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Done.

File: src/asr/asr_pytorch.py [hunk 1]
```diff
@@ -132,35 +132,49 @@ def update_core(self):
...
+        self.device = device
+        self.subsamping_factor = subsamping_factor
+        self.ignore_id = -1
+
+    def __call__(self, batch):
+        # batch should be located in list
+        assert len(batch) == 1
+        batch = batch[0]
+
+        # load acoustic features and target sequence of token ids
```
Comment (sw005320) [inline src/asr/asr_pytorch.py]: We can unify this data preprocessing part (before sending the data to GPU) with the one in `asr_chainer.py` as one function.

File: src/nets/e2e_asr.py [hunk 2]
```diff
@@ -36,15 +36,7 @@ def _subsamplex(x, n):
...
-# get output dim for latter BLSTM
-def _get_vgg2l_odim(idim, in_channel=3, out_channel=128):
-    idim = idim / in_channel
-    idim = np.ceil(np.array(idim, dtype=np.float32) / 2)  # 1st max pooling
-    idim = np.ceil(np.array(idim, dtype=np.float32) / 2)  # 2nd max pooling
-    idim = np.array(idim, dtype=np.int32)
-    return idim * out_channel  # numer of channels
-
-
+# TODO(kan-bayashi): no need to use linear tensor
```
Comment (sw005320) [inline src/nets/e2e_asr.py]: Can you replace linear_tensor() with the other function, and remove this function?
See https://github.com/espnet/espnet/issues/199

File: src/nets/e2e_asr_th.py [hunk 3]
```diff
@@ -1956,29 +1914,36 @@ def __init__(self, etype, idim, elayers, eunits, eprojs, subsample, dropout, in_
...
+            xs_pad, ilens = self.enc2(xs_pad, ilens)
         else:
             logging.error(
                 "Error: need to specify an appropriate encoder archtecture")
             sys.exit()
 
-        return xs, ilens
+        return xs_pad, ilens
+
+    def _get_vgg2l_odim(self, idim, in_channel=3, out_channel=128):
```
Comment (sw005320) [inline src/nets/e2e_asr_th.py]: The same function is used for pytorch and chainer. You can move this to asr_utils.py or whatever.

File: src/asr/asr_chainer.py [hunk 4]
```diff
@@ -244,7 +242,46 @@ def run(self):
...
+
+class CustomConverter(object):
+    """CUSTOM CONVERTER"""
+
+    def __init__(self, device, subsamping_factor=1):
+        self.device = device
+        self.subsamping_factor = subsamping_factor
+        self.xp = cuda.cupy if device != -1 else np
+
+    def __call__(self, batch, *args):
```
Comment (jnishi) [inline src/asr/asr_chainer.py]: Why `args` is required? It seems not to be used in this function and not to be given when this function is called anywhere.

File: src/asr/asr_pytorch.py (hunk 1 above)
Comment (jnishi) [reply src/asr/asr_pytorch.py]: In addition, it is useful to prefetch if loading from scp file becomes other function. cf. #340

File: src/asr/asr_chainer.py (hunk 4 above)
Comment (kan-bayashi) [reply src/asr/asr_chainer.py]: In `chainer.extensions.Eavaluator`, converter recieves 2 argments.
So I added but there may be more smart way :(

Comment (kan-bayashi) [reply src/asr/asr_chainer.py]: > In addition, it is useful to prefetch if loading from scp file becomes other function.
Is it necessary? I think just replacing converter is OK for your prefetcher.

File: src/nets/e2e_asr.py (hunk 2 above)
Comment (kan-bayashi) [reply src/nets/e2e_asr.py]: I will do.

File: src/nets/e2e_asr_th.py (hunk 3 above)
Comment (kan-bayashi) [reply src/nets/e2e_asr_th.py]: Integrated.

File: src/asr/asr_chainer.py (hunk 4 above)
Comment (jnishi) [reply src/asr/asr_chainer.py]: I see. Then it is better to send data to GPU in `converter` with `device` argument. It fits to Chainer's manner.

Comment (jnishi) [reply src/asr/asr_chainer.py]: If `device` argument and loading data to GPU is supported, `CustomWorker` and `setup_workers` may be removed.

File: src/nets/e2e_asr.py (hunk 2 above)
Comment (kan-bayashi) [reply src/nets/e2e_asr.py]: `n_batch_axes` option in `L.Linear` is not merged in v 4.3.1.
I will keep it.

File: src/asr/asr_chainer.py (hunk 4 above)
Comment (kan-bayashi) [reply src/asr/asr_chainer.py]: I will improve interface to be compatible chainer manner.

**Comment by kan-bayashi** [comment]
Done!

**Comment by sw005320** [comment]
great!
btw, you can further refactor the recog part in asr_chainer.py and asr_pytorch especially after it gets nbest.
We should do it after this PR. Now I just want to merge it because @jnishi can do prefetch implementation at the same time.

**Pull request merged**